### DISCUSSION TITLE
File Explorerer context menu shell extension options not appearing for un-hydrated placeholders

### DISCUSSION THREAD

**Discussion by FeralSquid** [microsoft/ProjFS-Managed-API#61] [Q&A]
Hello,

I've noticed that File Explorer tries to do some nice things to prevent unnecessary hydration of virtual files (not creating thumbnails by default, etc).

One thing it appears to do, though, is suppress shell extensions from adding context menu items to placeholders that aren't hydrated.

For example, I currently have a 7zip and perforce shell extensions installed, which should add menu items for any file I right-click.  These menu items do not show up, though, only if the file is an un-hydrated placeholder.

I was wondering if you knew of a way to control / configure this somehow?
Maybe whitelist certain extensions or something?

My users are annoyed by not getting the perforce menu options specifically.

Thanks!

**Comment by cgallred** [comment]
I don't know whether that's possible, though I suspect not. Since the Shell generally doesn't know what an extension might do I think they opt not to call extensions for any item in an enumeration that is marked with FILE_ATTRIBUTE_RECALL_ON_OPEN, lest they end up triggering a spurious hydration.

I've reached out to one of my Shell contacts to get more info.

**Reply by FeralSquid** [reply]
I was expecting that to likely be the answer, but asking to make sure.

Thanks for hunting the info down for me!

**Comment by cgallred** [comment]
Okay, the Shell folks got back to me. This isn't something the user of an extension can control, only the implementer:
> How are they implementing their context menu? Via IContextMenu? We only filter out those extension points. Static verbs are available.

To clarify, I asked for confirmation that the extension user can't control this, only the implementer. His response:
> Correct, we are trying to prevent unexpected hydrations from context menu extensions.

**Reply by FeralSquid** [reply]
Thanks for following up!

I'll poke perforce to see if we can convince them to change their shell extension implementation, or probably more likely make our own at some point.

**Reply by cgallred** [reply]
Sorry I couldn't give you an easy fix. :-)

One more piece of technical info for a context menu handler from the Shell team on how a shell extension can declare that they know how to work correctly with virtual files:

> Context menu handlers can opt into using cloud files; see the docs for the [DEFAULT_FOLDER_MENU_RESTRICTIONS enum](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/ne-shobjidl_core-default_folder_menu_restrictions#constants)

Name | Description
-- | --
... | ...
DFMR_OPTIN_HANDLERS_ONLY | 0x0040.   Opt-in to load handler for partial cloud files. Only used for handlers that   will not cause implicit hydration. Opt-in handlers must have the registry   value "ContextMenuOptIn" under `HKCR\CLSID<handler clsid>`
... | ...

> 7zip (and/or Perforce) will need to ensure they don’t cause files to be recalled and then set this option in their registration.

**Reply by FeralSquid** [reply]
great, thanks!
